source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+        // n-1].
+        i = first;
+        // Pattern to search for.
+        pattern = 0;
+        // Number of patterns found while looping through the list.
+        count = 0;
+
+        // Search for all patterns(values) within 1<<branch.
+        while (pattern < static_cast<uint32_t>(1 << branch)) {
+          pattern_found = false;
+          // Keep on looping until either all elements have been looked at
+          // stop looping when the pattern isn't matched anymore or we get the index out of bounds
+          while (i < first + n && static_cast<uint32_t>(extractBits<IpType, address_size>(
+                                      compute.prefix_, branch, ip_prefixes_[i].ip_)) == pattern) {
+            ++i;
+            pattern_found = true;
+          }
+          if (pattern_found) {
+            ++count;
+          }
+          ++pattern;
+        }
+        // Once the size of the branch(with the fill factor ratio) is able to contain all of the
+        // within the current range.
+      } while (count >= fill_factor_ * (1 << branch));
+
+      // Set the branching factor.
+      compute.branch_ = branch - 1;
+      return compute;
+    }
+
+    /**
+     * Recursively builds a tree that covers the base array from position 'first' to 'first+n'.
+     * @param prefix supplies the prefix to ignore when building the tree.
+     * @param first supplies the index into the ip_prefixes_ vector where to start building the
+     * trie_ from.
+     * @param n suppplies the number of entries.
+     * @param pos supplies the root for this trie..
+     * @param nextfree supplies the next available index in the trie_.
+     */
+    void buildRecursive(uint32_t prefix, uint32_t first, uint32_t n, uint32_t pos,
+                        uint32_t* nextfree) {
+      uint32_t p, address, bits;
+      uint32_t bit_pattern;
+      int k;
+
+      // Setting a leaf, the branch and skip are 0.
+      if (n == 1) {
+        trie_[pos] = first;
+        return;
+      }
+
+      ComputePair output = computeBranch(prefix, first, n);
+
+      address = *nextfree;
+      trie_[pos] =
+          setBranch(output.branch_) | setSkip(output.prefix_ - prefix) | setAddress(address);
+
+      // For the next part of the trie, the next available address is at address + 2^branching
+      // factor.
+      *nextfree += 1 << output.branch_;
+      p = first;
+
+      // Build the subtrees.
+      for (bit_pattern = 0; bit_pattern < static_cast<uint32_t>(1 << output.branch_);
+           ++bit_pattern) {
+
+        k = 0;
+        // k is finding the number of entries in the base vector that have the same bit pattern.
+        while (p + k < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[p + k].ip_)) == bit_pattern) {
+          ++k;
+        }
+
+        if (k == 0) {
+          if (p == first + n) {
```

Comment:
Do the unit tests currently cover both this and its "else" case?  I ask because the handling of the "k==0" situation (which requires interpolating the values for new child nodes that have to be created when the number of children for a node is less than 2^branch_) seems like one of the trickiest parts of the LC Trie construction.

Reply:
I double checked. The sample input provided in the paper covers all of the if/else cases in buildRecursive.